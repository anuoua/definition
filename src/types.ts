import * as varname from "varname";

export type KeyFormat = Exclude<keyof typeof varname, "default" | "split">;

export interface Resource {
  records: Record<string, any>[];
  utils: string;
  identify: string;
  out: string;
  valueField: string;
  keyFormat: KeyFormat;
  keyField: string;
  keyGen: (record: Record<string, any>) => string;
}

export interface ConfigOptions {
  dir: string;
  output: string;
  valueField?: string;
  keyField?: string;
  keyFormat?: KeyFormat;
  keyGen?: (record: Record<string, any>) => string;
}
