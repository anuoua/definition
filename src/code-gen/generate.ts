import ts from "typescript";
import { records } from "./sections/records";
import { enums } from "./sections/enums";
import { maps } from "./sections/maps";
import { imports } from "./sections/imports";
import { Resource } from "../types";

const { factory } = ts;

export const codeGen = (resource: Resource) => {
  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

  const importsAst = imports(resource);

  const recordsAst = records(resource);

  const enumsAst = enums(resource);

  const mapsAst = maps(resource);

  const astList = factory.createNodeArray(
    [importsAst, recordsAst, enumsAst, mapsAst].flat()
  );

  const sourceFile = factory.createSourceFile(
    astList,
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None
  );

  return printer.printFile(sourceFile);
};
