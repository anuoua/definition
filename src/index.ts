export * from "./types";

export { generate } from "./command-generate/action";
