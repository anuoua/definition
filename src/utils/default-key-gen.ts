import { KeyFormat } from "../types";
import * as varname from "varname";

export const defaultKeyGen = (
  record: Record<string, any>,
  keyField: string,
  keyFormat: KeyFormat
) => {
  if (!(keyField in record))
    throw new Error(`There is no \`${keyField}\` field in definition record`);
  return varname[keyFormat](record[keyField]);
};
