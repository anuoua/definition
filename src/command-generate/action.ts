import { cwd } from "process";
import { resolve, dirname, relative } from "path";
import { existsSync, writeFileSync, mkdirSync, cpSync } from "fs";
import { loadTsConfig } from "config-file-ts";
import { ConfigOptions } from "../types";
import { fromDirJsons } from "./from-dir-jsons";
import { codeGen } from "../code-gen/generate";

const resolvePath = (name: string) => resolve(cwd(), name);

const getConfig = async (config?: string) => {
  const paths = [
    resolvePath("definition.config.mjs"),
    resolvePath("definition.config.cjs"),
    resolvePath("definition.config.ts"),
  ];

  const configPath = config
    ? existsSync(resolvePath(config))
      ? config
      : undefined
    : await (async () => {
        for (let item of paths) {
          if (existsSync(item)) return item;
        }
      })();

  if (!configPath) throw new Error("Config file is not exist");

  if (configPath.endsWith(".mjs") || configPath.endsWith(".cjs")) {
    const model = await import(configPath);
    return (await model.default()) as ConfigOptions;
  } else {
    const config = loadTsConfig<() => Promise<ConfigOptions>>(configPath)!;
    return await config();
  }
};

export interface GenerateOptions {
  config?: string;
}

export const generate = async (options: GenerateOptions) => {
  const config = await getConfig(options.config);

  const resources = fromDirJsons(config);

  for (let resource of resources) {
    const outDir = dirname(resource.out);

    const code = codeGen(resource);

    if (!existsSync(outDir)) mkdirSync(outDir, { recursive: true });

    writeFileSync(resource.out, code);
  }

  cpSync("./utils", resolve(config.output, "utils"), {
    recursive: true,
  });
};
