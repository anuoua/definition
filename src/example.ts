import { vrMap } from "./utils/vr-map";
import { krMap } from "./utils/kr-map";
import { kvMap } from "./utils/kv-map";

export const TEST_records = [
  { key: "k-1", value: "value-1" },
  { key: "k-2", value: "value-3" },
] as const;
export type TEST_Records = typeof TEST_records;
export type TEST_Values = TEST_Records[number]["value"];
export enum TEST_Keys {
  K1 = "K1",
  K2 = "K2",
}
export const TEST_keys = [TEST_Keys.K1, TEST_Keys.K2] as const;
export const TEST_kr = krMap(TEST_records, TEST_keys);
export const TEST_vr = vrMap(TEST_records, TEST_keys);
export const TEST_kv = kvMap(TEST_records, TEST_keys);
