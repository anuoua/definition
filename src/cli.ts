import { program } from "commander";
import { GenerateOptions, generate } from "./command-generate/action";

program
  .command("generate")
  .alias("g")
  .option("-c, --config [path]", "Config file path")
  .action((options: GenerateOptions) => generate(options));

program.parse();

if (process.argv.length === 2) {
  process.argv.push("-h");
}
