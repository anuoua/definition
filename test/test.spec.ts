import { expect, it } from "bun:test";
import { codeGen } from "../src/code-gen/generate";
import { defaultKeyGen } from "../src/command-generate/default-key-gen";
import { generate } from "../src";

it("code generate", () => {
  const code = codeGen({
    identify: "TEST",
    valueField: "value",
    keyField: "key",
    keyFormat: "camelcase",
    keyGen: (record) => defaultKeyGen(record, "key", "camelcase"),
    out: "../",
    records: [
      { value: "hello", key: "hello-hello" },
      { value: "world", key: "world-world" },
    ],
    utils: "../",
  });

  expect(code).toMatchSnapshot();
});

it("from folder", async () => {
  await generate({ config: "./test/definition.config.ts" });
});
