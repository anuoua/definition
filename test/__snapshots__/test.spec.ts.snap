// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`code generate 1`] = `
"import { krMap, vrMap, kvMap } from "../";
export const TEST_records = [{ "value": "hello", "key": "hello-hello" }, { "value": "world", "key": "world-world" }] as const;
export type TEST_Records = typeof TEST_records;
export type TEST_Values = TEST_Records[number]["value"];
export enum TEST_Keys {
    HelloHello = "HelloHello",
    WorldWorld = "WorldWorld"
}
export const TEST_key_list = [TEST_Keys.HelloHello, TEST_Keys.WorldWorld] as const;
export const TEST_kr = krMap(TEST_records, TEST_key_list);
export const TEST_vr = vrMap(TEST_records, TEST_key_list, "value");
export const TEST_kv = kvMap(TEST_records, TEST_key_list, "value");
"
`;
